codex/blockexchange/engine/engine.nim

```diff
@@ -228,82 +244,99 @@ proc blockPresenceHandler*(
 
   let
     peerHave = peerCtx.peerHave
-    dontWantCids = peerHave.filterIt(it notin wantList)
+    dontWantCids = peerHave.filterIt(it notin ourWantList)
 
   if dontWantCids.len > 0:
     peerCtx.cleanPresence(dontWantCids)
 
-  let wantCids = wantList.filterIt(it in peerHave)
-
-  if wantCids.len > 0:
-    trace "Peer has blocks in our wantList", peer, wants = wantCids
-    await b.sendWantBlock(wantCids, peerCtx)
+  let ourWantCids = ourWantList.filter do(address: BlockAddress) -> bool:
+    if address in peerHave and not self.pendingBlocks.retriesExhausted(address) and
+        not self.pendingBlocks.isInFlight(address):
+      self.pendingBlocks.setInFlight(address, true)
+      self.pendingBlocks.decRetries(address)
```

Comment:
It is convenient to do it here, but I never like having side effects in `filter`... it also feels a bit out of context here when you read the code. We could add `setInFlight` that takes list of addresses. The same could be done for `self.pendingBlocks.decRetries`. We could also consider if  `self.pendingBlocks.setInFlight(address, true)` should be part of `sendWantBlock` (and then `self.self.pendingBlocks.setInFlight(address, false)` to be inside `sendWantHave` (but yeah, also said in other comment earlier that I like those two to be as focused as possible (still as long as they do not filter our blocks I am kind of fine with it).

Of course it is a bit faster to do it here, but , that could be a bit premature optimization....